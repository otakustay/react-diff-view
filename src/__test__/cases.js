/* In case you are looking for type of hunk, see src/Hunk/index.js */
import {parseDiff} from '..';

export const basic = `diff --git a/src/__test__/index.test.jsx b/src/__test__/index.test.jsx
index 643c2f0..7883597 100644
--- a/src/__test__/index.test.jsx
+++ b/src/__test__/index.test.jsx
@@ -21,3 +21,3 @@ describe('basic test', () => {
     test('App renders correctly', () => {
-        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();
+        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();
     });`;

export const basicHunk = parseDiff(basic)[0].hunks[0];

export const multiple = `diff --git a/src/index1.js b/src/index1.js
index e69de29..d00491f 100644
--- a/src/index1.js
+++ b/src/index1.js
@@ -0,0 +1 @@
+1
diff --git a/src/index.js b/src/index2.js
index d00491f..0cfbf08 100644
--- a/src/index2.js
+++ b/src/index2.js
@@ -1 +1 @@
-1
+2
diff --git a/src/index3.js b/src/index3.js
index 0cfbf08..e69de29 100644
--- a/src/index3.js
+++ b/src/index3.js
@@ -1 +0,0 @@
-2`;

export const multipleHunk = parseDiff(basic)[0].hunks[0];
