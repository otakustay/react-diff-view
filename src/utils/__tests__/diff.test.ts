// @ts-nocheck ignore: third-party
import {normalChange, insertChange, deleteChange} from '../../__tests__/changes.case';
import {basicHunk} from '../../__tests__/cases';
import {
    textLinesToHunk,
    insertHunk,
    expandFromRawCode,
    getCollapsedLinesCountBetween,
    expandCollapsedBlockBy,
    getChangeKey,
} from '..';

const noop = () => {/* noop */};

describe('textLinesToHunk', () => {
    test('basic', () => {
        const lines = [''];
        expect(textLinesToHunk(lines, 0, 0)).toMatchSnapshot();
    });
});

describe('insertHunk', () => {
    test('basic', () => {
        expect(insertHunk(
            [{changes: [{isInsert: true, lineNumber: 0, oldLineNumber: 0}]}],
            {changes: [{isInsert: true, lineNumber: 1, oldLineNumber: 1}]}
        )).toEqual(
            [{changes: [{isInsert: true, lineNumber: 1, oldLineNumber: 1}]}]
        );
    });
});

describe('expandFromRawCode', () => {
    test('basic', () => {
        const hunks = [basicHunk];
        expect(expandFromRawCode(hunks, '')).toMatchSnapshot();
    });
});

describe('getCollapsedLinesCountBetween', () => {
    test('empty', () => {
        const previousHunk = {};
        const nextHunk = {};
        expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(NaN);
    });

    test('basic', () => {
        const previousHunk = {oldStart: 1, oldLines: 2};
        const nextHunk = {oldStart: 10};
        expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(7);
    });

    test('minus number', () => {
        const previousHunk = {oldStart: 1, oldLines: 10};
        const nextHunk = {oldStart: 2};
        expect(getCollapsedLinesCountBetween(previousHunk, nextHunk)).toBe(-9);
    });

    test('no previousHunk', () => {
        const nextHunk = {oldStart: 2};
        expect(getCollapsedLinesCountBetween(null, nextHunk)).toBe(1);
    });

    test('throw when nextHunk is null', () => {
        expect(() => getCollapsedLinesCountBetween({}, null)).toThrow();
    });
});

describe('expandCollapsedBlockBy', () => {
    test('basic', () => {
        const hunks = [basicHunk];
        expect(expandCollapsedBlockBy(hunks, '', noop)).toMatchSnapshot();
    });

    test('normal hunk', () => {
        const hunks = [{
            content: '@@ -1,2 +1,2 @@',
            oldStart: 1,
            newStart: 1,
            oldLines: 2,
            newLines: 2,
            changes: [{
                content: 'iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj',
                type: 'normal',
                isNormal: true,
                oldLineNumber: 1,
                newLineNumber: 1,
            }, {
                content: 'dsds',
                type: 'delete',
                isDelete: true,
                lineNumber: 2,
            }, {
                content: 'dsdsds',
                type: 'insert',
                isInsert: true,
                lineNumber: 2,
            }],
            isPlain: false,
        }];
        const rawCode = 'iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj\ndsds';
        expect(expandCollapsedBlockBy(hunks, rawCode, noop)).toMatchSnapshot();
        expect(expandCollapsedBlockBy(hunks, rawCode.split('\n'), noop)).toMatchSnapshot();
        expect(expandFromRawCode(hunks, rawCode, 0, 10)).toMatchSnapshot();
    });
});

describe('getChangeKey', () => {
    test('throws when empty', () => {
        expect(() => getChangeKey()).toThrow();
    });

    test('normal change', () => {
        expect(getChangeKey(normalChange)).toBe('N0');
    });

    test('insert change', () => {
        expect(getChangeKey(insertChange)).toBe('I0');
    });

    test('delete change', () => {
        expect(getChangeKey(deleteChange)).toBe('D0');
    });
});
