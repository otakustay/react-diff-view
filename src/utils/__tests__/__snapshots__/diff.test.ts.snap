// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandCollapsedBlockBy basic 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "    test('App renders correctly', () => {",
        "isNormal": true,
        "newLineNumber": 21,
        "oldLineNumber": 21,
        "type": "normal",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
        "isDelete": true,
        "lineNumber": 22,
        "type": "delete",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
        "isInsert": true,
        "lineNumber": 22,
        "type": "insert",
      },
      Object {
        "content": "    });",
        "isNormal": true,
        "newLineNumber": 23,
        "oldLineNumber": 23,
        "type": "normal",
      },
    ],
    "content": "@@ -21,3 +21,3 @@ describe('basic test', () => {",
    "isPlain": false,
    "newLines": 3,
    "newStart": 21,
    "oldLines": 3,
    "oldStart": 21,
  },
]
`;

exports[`expandCollapsedBlockBy normal hunk 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
      Object {
        "content": "dsds",
        "isDelete": true,
        "lineNumber": 2,
        "type": "delete",
      },
      Object {
        "content": "dsdsds",
        "isInsert": true,
        "lineNumber": 2,
        "type": "insert",
      },
    ],
    "content": "@@ -1,2 +1,2 @@",
    "isPlain": false,
    "newLines": 2,
    "newStart": 1,
    "oldLines": 2,
    "oldStart": 1,
  },
]
`;

exports[`expandCollapsedBlockBy normal hunk 2`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
      Object {
        "content": "dsds",
        "isDelete": true,
        "lineNumber": 2,
        "type": "delete",
      },
      Object {
        "content": "dsdsds",
        "isInsert": true,
        "lineNumber": 2,
        "type": "insert",
      },
    ],
    "content": "@@ -1,2 +1,2 @@",
    "isPlain": false,
    "newLines": 2,
    "newStart": 1,
    "oldLines": 2,
    "oldStart": 1,
  },
]
`;

exports[`expandCollapsedBlockBy normal hunk 3`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 0,
        "oldLineNumber": 0,
        "type": "normal",
      },
      Object {
        "content": "iiiiiiiiiiiiiiiiiiiiii:WQiiiiiiiiiiiiejj",
        "isNormal": true,
        "newLineNumber": 1,
        "oldLineNumber": 1,
        "type": "normal",
      },
      Object {
        "content": "dsds",
        "isDelete": true,
        "lineNumber": 2,
        "type": "delete",
      },
      Object {
        "content": "dsdsds",
        "isInsert": true,
        "lineNumber": 2,
        "type": "insert",
      },
    ],
    "content": "@@ -0,3 +0,3",
    "isPlain": false,
    "newLines": 3,
    "newStart": 0,
    "oldLines": 3,
    "oldStart": 0,
  },
]
`;

exports[`expandFromRawCode basic 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "    test('App renders correctly', () => {",
        "isNormal": true,
        "newLineNumber": 21,
        "oldLineNumber": 21,
        "type": "normal",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();",
        "isDelete": true,
        "lineNumber": 22,
        "type": "delete",
      },
      Object {
        "content": "        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();",
        "isInsert": true,
        "lineNumber": 22,
        "type": "insert",
      },
      Object {
        "content": "    });",
        "isNormal": true,
        "newLineNumber": 23,
        "oldLineNumber": 23,
        "type": "normal",
      },
    ],
    "content": "@@ -21,3 +21,3 @@ describe('basic test', () => {",
    "isPlain": false,
    "newLines": 3,
    "newStart": 21,
    "oldLines": 3,
    "oldStart": 21,
  },
]
`;

exports[`textLinesToHunk basic 1`] = `
Object {
  "changes": Array [
    Object {
      "content": "",
      "isNormal": true,
      "newLineNumber": 0,
      "oldLineNumber": 0,
      "type": "normal",
    },
  ],
  "content": "@@ -0,1 +0,1",
  "isPlain": true,
  "newLines": 1,
  "newStart": 0,
  "oldLines": 1,
  "oldStart": 0,
}
`;
